usaspending_api/common/etl/spark.py

```diff
@@ -560,3 +564,176 @@ def create_ref_temp_views(spark: SparkSession, create_broker_views: bool = False
             spark.sql(sql_statement)
 
     logger.info(f"Created the reference views in the global_temp database")
+
+
+def write_csv_file(
+    spark: SparkSession,
+    df: DataFrame,
+    parts_dir: str,
+    max_rows_per_merged_file=EXCEL_ROW_LIMIT,
+    overwrite=True,
+    logger=None,
+) -> int:
+    """Write DataFrame data to CSV file parts.
+    Args:
+        spark: passed-in active SparkSession
+        df: the DataFrame wrapping the data source to be dumped to CSV.
+        parts_dir: Path to dir that will contain the outputted parts files from partitions
+        overwrite: Whether to replace the file CSV files if they already exist by that name
+        max_rows_per_merged_file: Suggestion to Spark of how many records to put in each written CSV file part,
+        if it will end up writing multiple files.
+        logger: The logger to use. If one note provided (e.g. to log to console or stdout) the underlying JVM-based
+            Logger will be extracted from the ``spark`` ``SparkSession`` and used as the logger.
+    Returns:
+        record count of the DataFrame that was used to populate the CSV file(s)
+    """
+    if not logger:
+        logger = get_jvm_logger(spark)
+    # Delete output data dir if it already exists
+    parts_dir_path = spark.sparkContext._jvm.org.apache.hadoop.fs.Path(parts_dir)
+    fs = parts_dir_path.getFileSystem(spark.sparkContext._jsc.hadoopConfiguration())
+    if fs.exists(parts_dir_path):
+        fs.delete(parts_dir_path, True)
+    start = time.time()
+    logger.info(f"Writing source data DataFrame to csv part files for file {parts_dir}...")
+    df_record_count = df.count()
+    df.write.options(
+        # NOTE: this is a suggestion, to be used by Spark if partitions yield multiple files
+        maxRecordsPerFile=max_rows_per_merged_file,
+    ).csv(
+        path=parts_dir,
+        header=False,
+        nullValue=None,
+        escape='"',  # " is used to escape the 'quote' character setting (which defaults to "). Escaped quote = ""
+        ignoreLeadingWhiteSpace=False,  # must set for CSV write, as it defaults to true
+        ignoreTrailingWhiteSpace=False,  # must set for CSV write, as it defaults to true
+        timestampFormat=CONFIG.SPARK_CSV_TIMEZONE_FORMAT,
+        mode="overwrite" if overwrite else "errorifexists",
+    )
+    logger.info(f"{parts_dir} contains {df_record_count:,} rows of data")
+    logger.info(f"Wrote source data DataFrame to csv part files in {(time.time() - start):3f}s")
+    return df_record_count
+
+
+def hadoop_copy_merge(
+    spark: SparkSession,
+    parts_dir: str,
+    header: str,
+    max_rows_per_merged_file=EXCEL_ROW_LIMIT,
+    logger=None,
+    file_format=".csv",
+) -> None:
+    """PySpark impl of Hadoop 2.x copyMerge() (deprecated in Hadoop 3.x)
+    It uses the underlying Hadoop FileSystem API to combine all part files under a directory
+    into a single specified destination file.
```

Comment:
Is this combining the part files into a single file or multiple files based on the `max_rows_per_merged_file` parameter? 

Also, it's not clear based on this initial comment and arg descriptions where this "specified destination file" is. If it really is multiple files, what will they be named, and where can I expect to find them?

Reply:
This is from Keith's implementation.

Reply:
> It uses the underlying Hadoop FileSystem API to combine all part files under a directory
    into a single specified destination file
    
    If I update "destination file" here to the parts parameter does it make more sense?

Reply:
I've gone ahead and updated the documentation. Let me know if it's clear enough now.

Reply:
Mostly just pointing out the inaccuracy of that first sentence. It's sometimes true, but in many cases it will not be. Is there a better description that would fit all cases?

Maybe something like:

> Merges files from a provided input directory and then redivides them into multiple files based on a provided maximum row count parameter.

Also, when a reader is trying to figure out where these files are going to end up (where and what they will be named), they might not think to check the `max_rows_per_merged_file` argument description, which right now only covers what they'll be named, not where they will end up. Because a list of file paths are not being returned by this method, this information should be obvious to the reader. Can you make this more clear somewhere in the Pydoc comment? Maybe a notes section at the end?

Reply:
> they might not think to check the max_rows_per_merged_file argument description,

They should, its the contract of the function. When someone is trying to understand how to work with a function they need to read all of it's documentation.

On the note of not returning file paths, I propose we update the method to return a list of merged files. I believe that will be necessary anyway when this job starts running queries with large data. I believe once this function returns a path to the files everything will be clear. Do you agree?

Reply:
Changes made based on teams meeting.